README.md
[XML-GrammarBase](http://www.shlomifish.org/open-source/projects/XML-Grammar/Base/)
code repository. Pull requests are welcome:

- [MetaCPAN Page](https://metacpan.org/release/XML-GrammarBase)


XML-GrammarBase/Changes
Revision history for XML-GrammarBase

0.2.10      2023-09-14
    - Better compatibility with bleadperl
        - Avoid string versions in "use" statements.

0.2.9       2022-04-20
    - Minor cleanups:
        - Remove unused scripts/bump-version.
        - Fix perltidy tests.
        - Use __END__ .

0.2.8       2020-10-17
    - dist.ini / weaver.ini / .tidyallrc / etc. cleanup
        - Move to @SHLOMIF
        - Inspired by Lady_Aleena.

0.2.7  2020-01-16
        - Remove checks for libxslt versions
            - https://rt.cpan.org/Ticket/Display.html?id=131155
            - Thanks to SREZIC and others
            - These checks were much maligned.

0.2.6  2018-05-03
        - Add prereq on MooX::late.
            - See https://rt.cpan.org/Public/Bug/Display.html?id=125264
            - Thanks to SREZIC .

0.2.5  2018-05-02
        - Convert to GitHub .

0.2.4  2018-05-02
        - Convert to Dist-Zilla

v0.2.3  2014-02-01
        - Add the separate LICENSE file (CPANTS Kwalitee).
        - Minimal version of perl (CPANTS).
        - Add t/style-trailing-space.t .

v0.2.2  2013-05-09
        - Bump the minimal version of XML-LibXML to 2.0017.
            - Done in order to avoid test failures with too low versions
            of libxml2 .
        - Add scripts/tag-release.pl .

v0.2.1  2013-01-23
        - Bump the minimal version of XML-LibXSLT to 1.80.
            - Done in order to avoid test failures.

v0.2.0  2013-01-20
        - Add support for XSLT parameters into the XSLT role.

v0.1.5  2013-01-18
        - More fixes to the tests and Unicode handling.
            - Test for properly emitting Unicode characters.

v0.1.4  2013-01-15
        - Add t/cpan-changes.t and tweaked "Changes" (= this file
        accordingly).

v0.1.3   2013-01-15
         - lib/XML/GrammarBase/Role/RelaxNG.pm : add the
         _calc_parser method that can be overrided for calculating and
         returning the XML::LibXML parser.

         - Fix module_base to only accept a distribution directory
         style (e.g: "XML-Grammar-MyGrammar" rather than the wrong
         "XML::Grammar::MyGrammar" which File::ShareDir's dist_dir did
         not like).
            - add a check for it in Moo using isa => ...

v0.1.2   2013-01-05
         - Add t/01libxml-basic.t for diag()s for libxml2 and XML::LibXML
         (so we'll know which versions of them are failing).
         - Add scripts/bump-version-number.pl .

v0.1.1   2013-01-03
         - Add keywords and resources to the META.yml / META.json.

v0.1.0   2013-01-03
         - Add the 'encoding' parameter for output_as_chars() vs.
         output_as_bytes().

v0.0.3   2012-12-29
         - Change a wrong synopsis - it was "use Any::Moose;" instead of
         "use MooX 'late';".

v0.0.2   2012-12-25
         - Convert to «use MooX::Role 'late';» instead of the previous
         incantation.
            - it broke the tests on many CPAN smokers and on a fresh perl
            installation.

v0.0.1  2012-12-03
        - First version, released on an unsuspecting world.



XML-GrammarBase/t/lib/Test/XML/Ordered.pm
package Test::XML::Ordered;

use strict;
use warnings;

use XML::LibXML::Reader;

use Test::More;

use base 'Exporter';

use vars '@EXPORT_OK';

@EXPORT_OK = (qw(is_xml_ordered));

sub new
{
    my $class = shift;
    my $self = {};

    bless $self, $class;

    $self->_init(@_);

    return $self;
}

sub _got
{
    return shift->{got_reader};
}

sub _expected
{
    return shift->{expected_reader};
}

sub _init
{
    my ($self, $args) = @_;

    $self->{got_reader} =
        XML::LibXML::Reader->new(@{$args->{got_params}});
    $self->{expected_reader} =
        XML::LibXML::Reader->new(@{$args->{expected_params}});

    $self->{diag_message} = $args->{diag_message};

    $self->{got_end} = 0;
    $self->{expected_end} = 0;

    return;
}

sub _got_end
{
    return shift->{got_end};
}

sub _expected_end
{
    return shift->{expected_end};
}

sub _read_got
{
    my $self = shift;

    if ($self->_got->read() <= 0)
    {
        $self->{got_end} = 1;
    }

    return;
}

sub _read_expected
{
    my $self = shift;

    if ($self->_expected->read() <= 0)
    {
        $self->{expected_end} = 1;
    }

    return;
}

sub _next_elem
{
    my $self = shift;

    $self->_read_got();
    $self->_read_expected();

    return;
}

sub _ns
{
    my $elem = shift;
    my $ns = $elem->namespaceURI();

    return defined($ns) ? $ns : "";
}

sub _compare_loop
{
    my $self = shift;

    my $calc_prob = sub {
        my $args = shift;

        if (!exists($args->{param}))
        {
            die "No 'param' specified.";
        }
        return
        {
            verdict => 0,
            param => $args->{param},
        }
    };

    NODE_LOOP:
    while ((!$self->_got_end()) && (!$self->_expected_end()))
    {
        my $type = $self->_got->nodeType();
        my $exp_type = $self->_expected->nodeType();

        if ($type == XML_READER_TYPE_SIGNIFICANT_WHITESPACE())
        {
            $self->_read_got();
            redo NODE_LOOP;
        }
        elsif ($exp_type == XML_READER_TYPE_SIGNIFICANT_WHITESPACE())
        {
            $self->_read_expected();
            redo NODE_LOOP;
        }
        elsif ($type != $exp_type)
        {
            return $calc_prob->({param => "nodeType"});
        }
        elsif ($type == XML_READER_TYPE_TEXT())
        {
            my $got_text = $self->_got->value();
            my $expected_text = $self->_expected->value();

            foreach my $t ($got_text, $expected_text)
            {
                $t =~ s{\A\s+}{}ms;
                $t =~ s{\s+\z}{}ms;
                $t =~ s{\s+}{ }ms;
            }
            if ($got_text ne $expected_text)
            {
                return $calc_prob->({param => "text"});
            }
        }
        elsif ($type == XML_READER_TYPE_ELEMENT())
        {
            if ($self->_got->name() ne $self->_expected->name())
            {
                return $calc_prob->({param => "element_name"});
            }
            if (_ns($self->_got) ne _ns($self->_expected))
            {
                return $calc_prob->({param => "mismatch_ns"});
            }
        }
    }
    continue
    {
        $self->_next_elem();
    }

    return { verdict => 1};
}

sub _get_diag_message
{
    my ($self, $status_struct) = @_;

    if ($status_struct->{param} eq "nodeType")
    {
        return
            "Different Node Type!\n"
            . "Got: " . $self->_got->nodeType() . " at line " . $self->_got->lineNumber()
            . "\n"
            . "Expected: " . $self->_expected->nodeType() . " at line " . $self->_expected->lineNumber()
            ;
    }
    elsif ($status_struct->{param} eq "text")
    {
        return
            "Texts differ: Got at " . $self->_got->lineNumber(). " with value <<@{[$self->_got->value()]}>> ; Expected at ". $self->_expected->lineNumber() . " with value <<@{[$self->_expected->value()]}>>.";
    }
    elsif ($status_struct->{param} eq "element_name")
    {
        return
            "Got name: " . $self->_got->name(). " at " . $self->_got->lineNumber() .
            " ; " .
            "Expected name: " . $self->_expected->name() . " at " .$self->_expected->lineNumber();
    }
    elsif ($status_struct->{param} eq "mismatch_ns")
    {
        return
            "Got Namespace: " . _ns($self->_got). " at " . $self->_got->lineNumber() .
            " ; " .
            "Expected Namespace: " . _ns($self->_expected) . " at " .$self->_expected->lineNumber();
    }

    else
    {
        die "Unknown param";
    }
}

sub compare
{
    local $Test::Builder::Level = $Test::Builder::Level+1;

    my $self = shift;

    $self->_next_elem();

    my $status_struct = $self->_compare_loop();
    my $verdict = $status_struct->{verdict};

    if (!$verdict)
    {
        diag($self->_get_diag_message($status_struct));
    }

    return ok($verdict, $self->{diag_message});
}

sub is_xml_ordered
{
    local $Test::Builder::Level = $Test::Builder::Level+1;

    my ($got_params, $expected_params, $message) = @_;

    my $comparator =
        Test::XML::Ordered->new(
            {
                got_params => $got_params,
                expected_params => $expected_params,
                diag_message => $message,
            }
        );

    return $comparator->compare();
}

1;



XML-GrammarBase/lib/XML/GrammarBase.pm
package XML::GrammarBase;

use warnings;
use strict;

use 5.008;

1;

__END__

=head1 NAME

XML::GrammarBase - Provide roles and base classes for processors of
specialized XML grammars.

=head1 SYNOPSIS

See the ones under the Role packages.

=head1 DESCRIPTION

XML::GrammarBase aims to be a convenient framework for easily providing
processors for XML grammars (such as those under the XML::Grammar namespace
- L<http://www.shlomifish.org/open-source/projects/XML-Grammar/> ).

It provides roles and base classes for facilitating writing those.

=head1 SEE ALSO

L<XML::GrammarBase::Role::RelaxNG> , L<XML::GrammarBase::Role::XSLT> .

=head1 FUNCTIONS

=head1 AUTHOR

Shlomi Fish, C<< <shlomif at cpan.org> >>

=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009 Shlomi Fish.

This program is distributed under the MIT (X11) License:
L<http://www.opensource.org/licenses/mit-license.php>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

=cut


XML-GrammarBase/lib/XML/GrammarBase/Role/XSLT.pm
package XML::GrammarBase::Role::XSLT;

use strict;
use warnings;

use Package::Variant
    importing => [ 'MooX::Role' => ['late'], ],
    subs      => [qw(has with)];

# use MooX 'late';

use XML::LibXML 2.0017;
use XML::LibXSLT 1.80;

use autodie;

sub make_variant
{
    my ( $class, $target_package, %args ) = @_;

    my $output_format = $args{output_format};

    with('XML::GrammarBase::Role::XSLT::Global');

    has "to_${output_format}_xslt_transform_basename" =>
        ( isa => 'Str', is => 'rw' );

    has "_to_${output_format}_stylesheet" => (
        isa     => "XML::LibXSLT::StylesheetWrapper",
        is      => 'rw',
        default => sub { return shift->_calc_stylesheet($output_format), },
        lazy    => 1,
    );

    return;
}

1;

__END__

=head1 NAME

XML::GrammarBase::Role::XSLT - a parameterized role for XSLT conversions.

=head1 SYNOPSIS

    package XML::Grammar::MyGrammar::ToOtherStuff;

    use MooX 'late';

    use XML::GrammarBase::Role::RelaxNG;
    use XML::GrammarBase::Role::XSLT;

    with ('XML::GrammarBase::Role::RelaxNG');
    with XSLT(output_format => 'html');
    with XSLT(output_format => 'docbook');

    has '+module_base' => (default => 'XML-Grammar-MyGrammar');
    has '+rng_schema_basename' => (default => 'my-grammar.rng');

    has '+to_html_xslt_transform_basename' => (default => 'mygrammar-xml-to-html.xslt');
    has '+to_docbook_xslt_transform_basename' => (default => 'mygrammar-xml-to-docbook.xslt');

    package main;

    my $xslt = XML::Grammar::MyGrammar::ToOtherStuff->new(
        data_dir => "/path/to/data-dir",
    );

    # Throws an exception on failure.
    my $as_html = $xslt->perform_xslt_translation(
        {
            output_format => 'html'
            source => {file => $input_filename, },
            output => "string",
        }
    );

=head1 PARAMATERS

=head2 output_format

A Perl identifier string identifying the format.

=head1 SLOTS

=head2 module_base

The basename of the module - used for dist dir.

=head2 data_dir

The data directory where the XML assets can be found (the RELAX NG schema, etc.)

=head2 rng_schema_basename

The Relax NG Schema basename.

=head2 to_${output_format}_xslt_transform_basename

The basename of the primary XSLT transform file. Should be overrided in
the constructor or using C<has '+to_html'>. For example:

    has '+to_html_xslt_transform_basename' => (default => 'fiction-xml-to-html.xslt');

=head1 METHODS

=head2 $self->rng_validate_dom($source_dom)

Validates the DOM ( $source_dom ) using the RELAX-NG schema.

=head2 $self->rng_validate_file($file_path)

Validates the file in $file_path using the RELAX-NG schema.

=head2 $self->rng_validate_string($xml_string)

Validates the XML in the $xml_string using the RELAX-NG schema.

=head2 $converter->perform_xslt_translation

=over 4

=item * my $final_source = $converter->perform_xslt_translation({output_format => $format, source => {file => $filename}, output => "string" })

=item * my $final_source = $converter->perform_xslt_translation({output_format => $format, source => {string_ref => \$buffer}, output => "string" })

=item * my $final_dom = $converter->perform_xslt_translation({output_format => $format, source => {file => $filename}, output => "dom" })

=item * my $final_dom = $converter->perform_xslt_translation({output_format => $format, source => {dom => $libxml_dom}, output => "dom" })

=item * my $final_dom = $converter->perform_xslt_translation({output_format => $format, source => {dom => $libxml_dom}, output => {file => $path_to_file,}, })

=item * my $final_dom = $converter->perform_xslt_translation({output_format => $format, source => {dom => $libxml_dom}, output => {fh => $filehandle,}, })

=back

This method does the actual conversion with the output format of
$format. The C<'source'> argument points to a hash-ref with
keys and values for the source. If C<'file'> is specified there it points to the
filename to translate (currently the only available source). If
C<'string_ref'> is specified it points to a reference to a string, with the
contents of the source XML. If C<'dom'> is specified then it points to an XML
DOM as parsed or constructed by XML::LibXML.

The C<'output'> key specifies the return value. A value of C<'string'> returns
the XML as a string, and a value of C<'dom'> returns the XML as an
L<XML::LibXML> DOM object. If it is a hash ref then it specifies a
C<'file'> or a C<'fh'> with a filepath or filehandle respectively.

An optional C<'encoding'> parameter determines if one should output the string
as C<'utf8'> (the deafult - using L<XML::LibXSLT>'s
output_as_chars()) or as C<'bytes'> - using its output_as_bytes() .

An optional 'xslt_params' parmater allows one to specify a hash of XSLT
parameters.

=cut

=head2 BUILD

L<Moo> constructor. For internal use.

=head2 make_variant

L<Package::Variant> constructor. For internal use.

=head1 AUTHOR

Shlomi Fish, C<< <shlomif at cpan.org> >>



=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009 Shlomi Fish.

This program is distributed under the MIT (X11) License:
L<http://www.opensource.org/licenses/mit-license.php>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

=cut


XML-GrammarBase/lib/XML/GrammarBase/Role/DataDir.pm
package XML::GrammarBase::Role::DataDir;

use strict;
use warnings;

=head1 NAME

XML::GrammarBase::Role::DataDir - provide the data_dir accessor.

=cut

use MooX::Role 'late';

use File::ShareDir qw(dist_dir);

my $_component_re = qr/[A-Za-z_]\w*/;

has 'module_base' => (
    isa => sub {
        my ($dist_name) = @_;
        if (
            not(   ( ref($dist_name) eq '' )
                && ( $dist_name =~ m/\A$_component_re(?:-$_component_re)*\z/ ) )
            )
        {
            die
"module_base must be a distribution string of components separated by dashes";
        }
    },
    ,
    is => 'rw'
);
has 'data_dir' => (
    isa     => 'Str',
    is      => 'rw',
    default => sub { return shift->_calc_default_data_dir(); },
    lazy    => 1,
);

sub _calc_default_data_dir
{
    my ($self) = @_;

    return dist_dir( $self->module_base() );
}

sub _undefize
{
    my $class = shift;
    my $v     = shift;

    return defined($v) ? $v : "(undef)";
}

sub dist_path
{
    my ( $self, $basename ) = @_;

    return File::Spec->catfile( $self->data_dir, $basename );
}

sub dist_path_slot
{
    my ( $self, $slot ) = @_;

    return $self->dist_path( $self->$slot() );
}

1;

__END__

=head1 SYNOPSIS

    package MyClass::WithDataDir;

    use MooX 'late';

    with ('XML::GrammarBase::Role::DataDir');

    has '+module_base' => (default => 'XML-Grammar-MyGrammar');

    package main;

    my $obj = MyClass::WithDataDir->new(
        data_dir => "/path/to/data-dir",
    );

=head1 SLOTS

=head2 module_base

The basename of the distribution - used for dist dir.

=head2 data_dir

The data directory where the XML assets can be found (the RELAX NG schema, etc.)

=head1 METHODS

=head2 $self->dist_path($basename)

Returns the $basename relative to data_dir().

Utility method.

=head2 $self->dist_path_slot($slot)

Returns the basename of $self->$slot() relative to data_dir().

Utility method.

=head1 AUTHOR

Shlomi Fish, C<< <shlomif at cpan.org> >>



=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009 Shlomi Fish.

This program is distributed under the MIT (X11) License:
L<http://www.opensource.org/licenses/mit-license.php>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

=cut


XML-GrammarBase/lib/XML/GrammarBase/Role/RelaxNG.pm
package XML::GrammarBase::Role::RelaxNG;

use strict;
use warnings;

=head1 NAME

XML::GrammarBase::Role::RelaxNG - base class for a RelaxNG validator

=cut

use MooX::Role 'late';

use File::ShareDir qw(dist_dir);
use XML::LibXML 2.0017;

with('XML::GrammarBase::Role::DataDir');

has 'rng_schema_basename' => ( isa => 'Str', is => 'rw' );
has '_rng' => (
    isa     => 'XML::LibXML::RelaxNG',
    is      => 'rw',
    default => sub { return shift->_calc_default_rng_schema; },
    lazy    => 1,
);

sub _calc_default_rng_schema
{
    my ($self) = @_;

    my $rngschema =
        XML::LibXML::RelaxNG->new(
        location => $self->dist_path_slot('rng_schema_basename'), );

    return $rngschema;
}

sub rng_validate_dom
{
    my ( $self, $source_dom ) = @_;

    my $ret_code;

    eval { $ret_code = $self->_rng()->validate($source_dom); };

    if ( defined($ret_code) && ( $ret_code == 0 ) )
    {
        # It's OK.
    }
    else
    {
        confess "RelaxNG validation failed [\$ret_code == "
            . $self->_undefize($ret_code)
            . " ; $@]";
    }

    return;
}

sub _calc_parser
{
    my ($self) = @_;

    my $xml_parser = XML::LibXML->new();

    $xml_parser->validation(0);
    $xml_parser->load_ext_dtd(0);
    $xml_parser->no_network(1);

    return $xml_parser;
}

sub rng_validate_file
{
    my ( $self, $filename ) = @_;

    my $dom = $self->_calc_parser()->parse_file($filename);

    return $self->rng_validate_dom($dom);
}

sub rng_validate_string
{
    my ( $self, $xml_string ) = @_;

    my $dom = $self->_calc_parser()->parse_string($xml_string);

    return $self->rng_validate_dom($dom);
}

1;

__END__

=head1 SYNOPSIS

    package XML::Grammar::MyGrammar::RelaxNG::Validate;

    use MooX 'late';

    with ('XML::GrammarBase::Role::RelaxNG');

    has '+module_base' => (default => 'XML::Grammar::MyGrammar');
    has '+rng_schema_basename' => (default => 'my-grammar.rng');

    package main;

    my $rnger = XML::Grammar::MyGrammar::RelaxNG::Validate->new(
        data_dir => "/path/to/data-dir",
    );

    # Throws an exception on failure.
    $rnger->rng_validate_file("/different-path-to-xml-file.xml");

=head1 SLOTS

=head2 module_base

The basename of the module - used for dist dir.

=head2 data_dir

The data directory where the XML assets can be found (the RELAX NG schema, etc.)

=head2 rng_schema_basename

The Relax NG Schema basename.

=head1 METHODS

=head2 $self->rng_validate_dom($source_dom)

Validates the DOM ( $source_dom ) using the RELAX-NG schema.

=head2 $self->rng_validate_file($file_path)

Validates the file in $file_path using the RELAX-NG schema.

=head2 $self->rng_validate_string($xml_string)

Validates the XML in the $xml_string using the RELAX-NG schema.

=head2 $self->dist_path($basename)

Returns the $basename relative to data_dir().

Utility method.

=head2 $self->dist_path_slot($slot)

Returns the basename of $self->$slot() relative to data_dir().

Utility method.

=head2 BUILD

L<Any::Moose> constructor. For internal use.

=head1 AUTHOR

Shlomi Fish, C<< <shlomif at cpan.org> >>



=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009 Shlomi Fish.

This program is distributed under the MIT (X11) License:
L<http://www.opensource.org/licenses/mit-license.php>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

=cut


XML-GrammarBase/lib/XML/GrammarBase/Role/XSLT/Global.pm
package XML::GrammarBase::Role::XSLT::Global;

use strict;
use warnings;

=head1 NAME

XML::GrammarBase::Role::XSLT::Global - a base, non-parameterised, role for an XSLT converter.

=cut

use MooX::Role 'late';

use XML::LibXML 2.0017;
use XML::LibXSLT 1.80;

use autodie;

with('XML::GrammarBase::Role::RelaxNG');

has '_xml_parser' => (
    isa     => "XML::LibXML",
    is      => 'rw',
    default => sub { return XML::LibXML->new; },
    lazy    => 1,
);

has '_xslt_parser' => (
    isa     => "XML::LibXSLT",
    is      => 'rw',
    default => sub { return XML::LibXSLT->new; },
    lazy    => 1,
);

sub _calc_stylesheet
{
    my ( $self, $output_format ) = @_;

    my $style_doc =
        $self->_xml_parser()
        ->parse_file(
        $self->dist_path_slot("to_${output_format}_xslt_transform_basename"), );

    return $self->_xslt_parser->parse_stylesheet($style_doc);
}

sub _calc_and_ret_dom_without_validate
{
    my $self = shift;
    my $args = shift;

    my $source = $args->{source};

    return
          exists( $source->{'dom'} ) ? $source->{'dom'}
        : exists( $source->{'string_ref'} )
        ? $self->_xml_parser()->parse_string( ${ $source->{'string_ref'} } )
        : $self->_xml_parser()->parse_file( $source->{'file'} );
}

sub _get_dom_from_source
{
    my $self = shift;
    my $args = shift;

    my $source_dom = $self->_calc_and_ret_dom_without_validate($args);

    my $ret_code;

    eval { $ret_code = $self->_rng()->validate($source_dom); };

    if ( defined($ret_code) && ( $ret_code == 0 ) )
    {
        # It's OK.
    }
    else
    {
        confess "RelaxNG validation failed [\$ret_code == "
            . $self->_undefize($ret_code)
            . " ; $@]";
    }

    return $source_dom;
}

sub perform_xslt_translation
{
    my ( $self, $args ) = @_;

    my $output_format = $args->{output_format};
    my $encoding      = ( $args->{encoding} || 'utf8' );

    my $source_dom = $self->_get_dom_from_source($args);

    my $stylesheet_method = "_to_${output_format}_stylesheet";
    my $stylesheet        = $self->$stylesheet_method();

    my $medium = $args->{output};

    my $is_string = ( $medium eq 'string' );
    my $is_dom    = ( $medium eq 'dom' );

    my $xslt_params = $args->{xslt_params} || {};

    if ( $is_string or $is_dom )
    {
        my $results = $stylesheet->transform( $source_dom, %$xslt_params );

        return
              $is_dom                  ? $results
            : ( $encoding eq 'bytes' ) ? $stylesheet->output_as_bytes($results)
            :                            $stylesheet->output_as_chars($results);
    }
    elsif ( ref($medium) eq 'HASH' )
    {
        if ( exists( $medium->{'file'} ) )
        {
            open my $out, '>', $medium->{'file'};
            $self->perform_xslt_translation(
                {
                    %$args,
                    output   => { fh => $out, },
                    encoding => 'bytes',
                }
            );
            close($out);
            return;
        }
        if ( exists( $medium->{'fh'} ) )
        {
            print { $medium->{'fh'} } $self->perform_xslt_translation(
                {
                    %$args, output => "string",
                }
            );
            return;
        }
    }

    confess "Unknown medium";
}

1;

__END__

=head1 SYNOPSIS

    package XML::Grammar::MyGrammar::RelaxNG::Validate;

    use Moo;

    with ('XML::GrammarBase::Role::XSLT::Global');

=head1 DESCRIPTION

This is a utility global (non-variant) role used by
L<XML::GrammarBase::Role::XSLT> . For internal use.

=head1 METHODS

=head2 $self->perform_xslt_translation(...)

See L<XML::GrammarBase::Role::XSLT> .

=head1 AUTHOR

Shlomi Fish, C<< <shlomif at cpan.org> >>



=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009 Shlomi Fish.

This program is distributed under the MIT (X11) License:
L<http://www.opensource.org/licenses/mit-license.php>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

=cut


XML-GrammarBase/Build.PL
use strict;
use warnings;

use lib "./inc";

use Test::Run::Builder;

my $builder = Test::Run::Builder->new(
    module_name         => 'XML::GrammarBase',
    license             => 'mit',
    dist_author         => q{Shlomi Fish <shlomif@cpan.org>},
    dist_version_from   => 'lib/XML/GrammarBase.pm',
    configure_requires => {
        'Module::Build' => '0.36',
    },
    requires => {
        'autodie' => 0,
        'File::ShareDir' => 0,
        'Moo::Role' => 0,
        'MooX' => 0,
        'MooX::Role' => 0,
        'MooX::late' => 0,
        'MooX::Types::MooseLike' => 0,
        'Package::Variant' => 0,
        'strict' => 0,
        'warnings' => 0,
        'XML::LibXML' => '2.0017',
        'XML::LibXSLT' => '1.80',
    },
    build_requires => {
        'File::Temp' => 0,
        'Test::More' => 0,
        'Test::XML::Ordered' => 0,
    },
    add_to_cleanup      => [ 'XML-GrammarBase-*' ],
    meta_merge =>
    {
        resources =>
        {
            repository => "http://bitbucket.org/shlomif/perl-xml-grammarbase",
            homepage => "http://www.shlomifish.org/open-source/projects/XML-Grammar/Base/",
        },
        keywords =>
        [
            "abstraction",
            "base",
            "grammar",
            "moo",
            "moose",
            "relaxng",
            "rng",
            "role",
            "roles",
            "xml",
            "xslt",
        ],
    }
);

$builder->create_build_script();
